Add support for ordered @Inject in class hierarchies

TWiStErRob opened the pull request:
Re-work of https://github.com/JakeWharton/SdkSearch/pull/144 on top of current master addressing code review comments as well, and moving test as an integration test.

JakeWharton commented on integration-tests/src/main/java/com/example/MemberInjectionOrder.java:
```diff
@@ -0,0 +1,54 @@
+  void inject(SubType target);
+
+  abstract class Base {
+    List<String> calls = new LinkedList<>();
```
```suggestion
    List<String> calls = new ArrayList<>();
```

JakeWharton commented on integration-tests/src/test/java/com/example/IntegrationTest.java:
```diff
@@ -142,6 +142,27 @@
+    MemberInjectionOrder component = backend.create(MemberInjectionOrder.class);
+    MemberInjectionOrder.SubType target = new MemberInjectionOrder.SubType();
+    component.inject(target);
+    target.calls.add("injected: " + target);
```
I don't think you need this or the booleans to track method calls. The logs already contain that information. All we need to see is constructor, base method, and sub method which will inform the field injection order based on the output.

TWiStErRob replied:
good point about booleans, came over with migration; re "injected", wanted to see `subCalled` being true, but that's just testing the fixture now that I look at it fresh

JakeWharton commented on integration-tests/src/test/java/com/example/IntegrationTest.java:
```diff
@@ -142,6 +142,27 @@
   }
 
+  /**
+   * [@Inject] Constructors are injected first, followed by fields, and then methods. Fields and methods in
```
`{@code Inject @Inject}`, although I don't think Javadoc is useful on tests. I'd rather see this as a line comment inside near the assertion.

TWiStErRob commented:
Rebased and fixed comments, see [force-push](https://github.com/JakeWharton/dagger-reflect/compare/83e8cfece985376ac712a391198c13423e44d623..054d2091f1da69de5e5b1c78dcff411476a43fda) for details, again `ArrayDeque` because `addFirst` and errorprone.
